Test failures on CI spawning threads should skip tests instead

There was a CI test failure([here](https://github.com/numpy/numpy/actions/runs/14537275679/job/40788394181?pr=28768)):

<details>
<summary>failure message</summary>

```

=================================== FAILURES ===================================
________________________ test_parallel_ufunc_execution _________________________

    def test_parallel_ufunc_execution():
        # if the loop data cache or dispatch cache are not thread-safe
        # computing ufuncs simultaneously in multiple threads leads
        # to a data race that causes crashes or spurious exceptions
        def func():
            arr = np.random.random((25,))
            np.isnan(arr)
    
>       run_threaded(func, 500)

func       = <function test_parallel_ufunc_execution.<locals>.func at 0xd53ddda8>

numpy/_core/tests/test_multithreading.py:33: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
numpy/testing/_private/utils.py:2743: in run_threaded
    futures.append(tpe.submit(*arg))
        _          = 0
        all_args   = [(<function test_parallel_ufunc_execution.<locals>.func at 0xd53ddda8>,), (<function test_parallel_ufunc_execution.<lo...execution.<locals>.func at 0xd53ddda8>,), (<function test_parallel_ufunc_execution.<locals>.func at 0xd53ddda8>,), ...]
        arg        = (<function test_parallel_ufunc_execution.<locals>.func at 0xd53ddda8>,)
        args       = []
        func       = <function test_parallel_ufunc_execution.<locals>.func at 0xd53ddda8>
        futures    = [<Future at 0xd53fb118 state=finished returned NoneType>, <Future at 0xd53fb1a8 state=finished returned NoneType>, <Fu... <Future at 0xd53ef040 state=finished returned NoneType>, <Future at 0xd53ef598 state=finished returned NoneType>, ...]
        max_workers = 500
        outer_iterations = 1
        pass_barrier = False
        pass_count = False
        prepare_args = None
        tpe        = <concurrent.futures.thread.ThreadPoolExecutor object at 0xd53ba400>
/usr/lib/python3.10/concurrent/futures/thread.py:176: in submit
    self._adjust_thread_count()
        args       = ()
        f          = <Future at 0xd52bb2c8 state=finished returned NoneType>
        fn         = <function test_parallel_ufunc_execution.<locals>.func at 0xd53ddda8>
        kwargs     = {}
        self       = <concurrent.futures.thread.ThreadPoolExecutor object at 0xd53ba400>
        w          = <concurrent.futures.thread._WorkItem object at 0xd52bb2f8>
/usr/lib/python3.10/concurrent/futures/thread.py:199: in _adjust_thread_count
    t.start()
        num_threads = 225
        self       = <concurrent.futures.thread.ThreadPoolExecutor object at 0xd53ba400>
        t          = <Thread(ThreadPoolExecutor-3_225, initial)>
        thread_name = 'ThreadPoolExecutor-3_225'
        weakref_cb = <function ThreadPoolExecutor._adjust_thread_count.<locals>.weakref_cb at 0xd52a2ad8>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Thread(ThreadPoolExecutor-3_225, initial)>

    def start(self):
        """Start the thread's activity.
    
        It must be called at most once per thread object. It arranges for the
        object's run() method to be invoked in a separate thread of control.
    
        This method will raise a RuntimeError if called more than once on the
        same thread object.
    
        """
        if not self._initialized:
            raise RuntimeError("thread.__init__() not called")
    
        if self._started.is_set():
            raise RuntimeError("threads can only be started once")
    
        with _active_limbo_lock:
            _limbo[self] = self
        try:
>           _start_new_thread(self._bootstrap, ())
E           RuntimeError: can't start new thread

self       = <Thread(ThreadPoolExecutor-3_225, initial)>

/usr/lib/python3.10/threading.py:[935](https://github.com/numpy/numpy/actions/runs/14537275679/job/40788394181?pr=28768#step:6:936): RuntimeError
============================= slowest 10 durations =============================
45.55s call     build-install/usr/lib/python3/dist-packages/numpy/_core/tests/test_mem_overlap.py::test_diophantine_fuzz
10.87s call     build-install/usr/lib/python3/dist-packages/numpy/lib/tests/test_io.py::TestSaveTxt::test_large_zip
6.64s call     build-install/usr/lib/python3/dist-packages/numpy/_core/tests/test_mem_overlap.py::test_may_share_memory_easy_fuzz
5.85s call     build-install/usr/lib/python3/dist-packages/numpy/lib/tests/test_format.py::test_memmap_roundtrip
5.68s call     build-install/usr/lib/python3/dist-packages/numpy/_core/tests/test_multiarray.py::TestBool::test_count_nonzero_all
5.56s call     build-install/usr/lib/python3/dist-packages/numpy/_core/tests/test_mem_overlap.py::test_may_share_memory_harder_fuzz
5.53s call     build-install/usr/lib/python3/dist-packages/numpy/linalg/tests/test_regression.py::TestRegression::test_openblas_threading
4.97s call     build-install/usr/lib/python3/dist-packages/numpy/random/tests/test_extending.py::test_cython
4.73s call     build-install/usr/lib/python3/dist-packages/numpy/_core/tests/test_scalarmath.py::test_array_scalar_ufunc_equivalence
4.07s call     build-install/usr/lib/python3/dist-packages/numpy/linalg/tests/test_linalg.py::TestCond::test_generalized_sq_cases
=========================== short test summary info ============================
FAILED numpy/_core/tests/test_multithreading.py::test_parallel_ufunc_execution - RuntimeError: can't start new thread
= 1 failed, 46976 passed, 2170 skipped, 37 xfailed, 4 xpassed in 282.89s (0:04:42) =
Error: Process completed with exit code 1.
```
</details>

When an exception like this happens, we should catch the exception and skip the test.

_Originally posted by @Schefflera-Arboricola in https://github.com/numpy/numpy/issues/28768#issuecomment-2815830537_

--- a/numpy/testing/_private/utils.py
+++ b/numpy/testing/_private/utils.py
@@ -2741,6 +2741,11 @@ def run_threaded(func, max_workers=8, pass_count=False,
                 futures = []
                 for arg in all_args:
                     futures.append(tpe.submit(*arg))
+            except RuntimeError as e:
+                import pytest
+                pytest.skip(f"Spawning {max_workers} threads failed with "
+                            f"error {e!r} (likely due to resource limits on the "
+                            "system running the tests)")
             finally:
                 if len(futures) < max_workers and pass_barrier:
                     barrier.abort()

